### PR TITLE
Fix the TinyMCE init array

### LINKED ISSUE #10966: Fix the classic block TinyMCE init array
The TinyMCE init array cannot be JSON encoded safely like in `$scripts->localize()`. Some setting values can be (js) functions, other may be passed as objects or arrays in the `tiny_mce_before_init` filter. It has to be converted from php to js "by hand" and outputted as raw js from php.

### PR DESCRIPTION
**Pull request by azaozz**
closes #10966, #11147, #7942, #8670

- Added "manual" convert from php to js, same as in class-wp-editor.php.
- Reformatted the code there so it's more readable.

### CODE REVIEW

File: lib/client-assets.php [hunk 1]
```diff
@@ -536,104 +536,110 @@ function gutenberg_register_scripts_and_styles() {
...
+		'wpautoresize',
+		'wpeditimage',
+		'wpemoji',
+		'wpgallery',
+		'wplink',
+		'wpdialogs',
+		'wptextpattern',
+		'wpview',
+	);
+	$tinymce_plugins = apply_filters( 'tiny_mce_plugins', $tinymce_plugins, 'classic-block' );
```
Comment (iseulde) [inline lib/client-assets.php]: Why shouldn't we pass `editor` if we want it to be as close as possible like the old editor?

**Comment by iseulde** [comment]
> Added "manual" convert from php to js, same as in class-wp-editor.php.

Is it not possible to compose the right PHP object, then JSON encode?

File: lib/client-assets.php (hunk 1 above)
Comment (azaozz) [reply lib/client-assets.php]: That is (supposedly) the editor id and is intended to identify the TinyMCE instance when more than one on the same page. The classic editor passes `content` there, but thinking it's no good to reuse that as the classic block editor instance is quite different and plugins should be able to identify it from php.

Passing `editor` is a bit too generic :)

**Comment by azaozz** [comment]
No, because that breaks some of the functionality. The TinyMCE init is supposed to be written in (pure) js. It is in our (WordPress) implementation that we set it as php and then "convert" to js before output.

Several settings in the init object accept js functions. If it is JSON encoded, that breaks. Example: `init_instance_callback`. To test can use the mini WP plugin from here: https://github.com/WordPress/gutenberg/issues/10966#issuecomment-432378456.

Currently when using that plugin we get this in the console:
```
Uncaught TypeError: Cannot read property 'apply' of undefined
    at Editor.execCallback (tinymce.js?ver=4800-20180716)
    at initEditor (tinymce.js?ver=4800-20180716)
    .....
```
Happens because the init js is outputted by `wp_localize_script()` which in meant only for strings and JSON encodes the output.

With this pr we get:
```
Editor: editor-7ffd9aaf-ae66-47f1-a160-1820fe202478 is now initialized.
```
i.e. works as expected.

**Comment by azaozz** [comment]
Also, this pr adds another (custom) option to the TinyMCE init object: `'classic_block_editor' => true,`. It would be nice to be able to identify the TinyMCE instance from js based plugins. Will add similar option to the Classic editor's instance, probably `content_editor` => true,` (or maybe that should be `classic_editor`?).

**Comment by azaozz** [comment]
Heh? Tests failing because... the code is more readable? :)
```604 | WARNING | [x] Array double arrow not aligned correctly;
     |         |     expected 13 space(s) between "'toolbar1'" and
     |         |     double arrow, but found 1.
```
Who added that (very questionable) rule there? This seems contrary to the WP coding standards. So when defining arrays they should look like:
```
'long-array-key-that-would-make-for-a-long-line-and-there-is-no-way-to-wrap-it-up' => 'long-array-key-value-that-is-not-easy-to-read-as-it-makes-the-line-too-long',
'key'                                                                              => 'value',
```

The spacing looks so "borked" :) How is that easy to read? LOL

**Comment by danielbachhuber** [comment]
Note: this doesn't quite solve https://github.com/WordPress/gutenberg/issues/11147#issuecomment-434362555

**Comment by azaozz** [comment]
Also opened https://core.trac.wordpress.org/ticket/45221 for core. 

This is a regression and a blocker for WP 5.0.

**Comment by azaozz** [comment]
Fixed in core https://core.trac.wordpress.org/changeset/43867. The fix needs to be merged here too.

**Comment by youknowriad** [comment]
Is this ready to land? Can we rebase and merge @iseulde @azaozz ?

**Review by youknowriad — APPROVED** [review]
LGTM 👍

**Comment by youknowriad** [comment]
These changes are already in Core, so no need to backport.

**Comment by azaozz** [comment]
Ughhh, PHPCS forcing ugly code formatting :(

**Pull request merged**